### PR TITLE
[TfLite][GPU/CPU] Added GroupNormalization CPU/GPU kernel 

### PR DESCRIPTION
**Pull request by mgokulkrish**
Added custom layer GroupNormalization in TfLite

-> Needed for executing StableDiffusion Model on GPU in tflite.
```
from keras_cv.models.stable_diffusion.diffusion_model import DiffusionModel
from keras_cv.models.stable_diffusion.decoder import Decoder
```
-> limitations of GPU kernel
        - defined for axis=-1
        - scale and centre should be true
        - channels/groups >= 4

-> single layer group norm accuracy on s23 ultra - mean absolute error of e^-5 on fp32 execution for gpu
   for the configurations present in stable diffusion.

### CODE REVIEW

**Comment by mgokulkrish** [comment]
> Hi @mgokulkrish Can you please sign CLA. Thank you!

Hi @gbaned, I have already signed CLA in this PR, should i do it again ? #60749

File: tensorflow/lite/delegates/gpu/common/tasks/group_normalization_test_util.cc [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+#include "tensorflow/lite/delegates/gpu/common/status.h"
+#include "tensorflow/lite/delegates/gpu/common/task/testing_util.h"
+#include "tensorflow/lite/delegates/gpu/common/tasks/group_normalization.h"
+
+namespace tflite {
+namespace gpu {
+
+absl::Status GroupNormalizationTest(TestExecutionEnvironment* env,
+                              bool constant_idx) {
+  TensorFloat32 src_tensor;
```
Comment (KirovClassBattlecruiser) [inline tensorflow/lite/delegates/gpu/common/tasks/group_normalization_test_util.cc:31]: Could define this whole bunch of variables in a shorter way

File: tensorflow/lite/delegates/gpu/common/transformations/model_transformations.cc [hunk 2]
```diff
@@ -66,7 +67,9 @@ bool ApplyGeneralTransformations(ModelTransformer* transformer) {
          transformer->Apply("merge_add_with_convolution",
                             NewMergeAddWithConvolution().get()) &&
          transformer->Apply("merge_mul_with_convolution",
-                            NewMergeMulWithConvolution().get());
```
Comment (KirovClassBattlecruiser) [inline tensorflow/lite/delegates/gpu/common/transformations/model_transformations.cc:69]: Excess line

**Review by ghost — APPROVED** [review]
Good code overrall, just one excess line and a couple of variables declared in a non optimal way

File: tensorflow/lite/delegates/gpu/common/transformations/model_transformations.cc (hunk 2 above)
Comment (mgokulkrish) [reply tensorflow/lite/delegates/gpu/common/transformations/model_transformations.cc:69]: These are not part of my changes, its already present in the TensorFlow Lite, plus it is not repetition. The above one is for Add and the marked one is for Mul

File: tensorflow/lite/delegates/gpu/common/tasks/group_normalization_test_util.cc (hunk 1 above)
Comment (mgokulkrish) [reply tensorflow/lite/delegates/gpu/common/tasks/group_normalization_test_util.cc:31]: These tensor names are standard across all the tests, so it should be fine. The sample reference test :- 
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/delegates/gpu/common/tasks/pooling_test_util.cc

**Comment by mgokulkrish** [comment]
Hi @yijie-yang, could you please review this PR. Doing open-source contribution for the first time, would love to learn from the mistakes and improve.

**Comment by gbaned** [comment]
> > Hi @mgokulkrish Can you please sign CLA. Thank you!
> 
> Hi @gbaned, I have already signed CLA in this PR, should i do it again ? #60749

Hi @mgokulkrish Sorry for the delay, It shows cla check failed. Can you please make sure to use same GitHub username and email-id associated with it.

**Comment by mgokulkrish** [comment]
Hi @gbaned, I think the cla now is approved.

**Comment by gbaned** [comment]
> Hi @gbaned, I think the cla now is approved.

Hi @mgokulkrish Yes, it is good now. Thank you!

**Comment by mgokulkrish** [comment]
> Come on, somebody needs check the this merging problem. Btw @mgokulkrish did u have interest to make a repo to show how to use GPU to acclerate sd model on Android? sounds very interest for me

Hi @fuerpy, sorry I do not have open-source setup for it.

**Comment by ZTMIDGO** [comment]
@mgokulkrish hello，does it mean that this model can also use the GPU

https://huggingface.co/keras-sd/diffusion-model-tflite/tree/main

**Review by JunyoungLim — COMMENTED** [review]
Hi @mgokulkrish, the recommended way of adding custom op (only supported in C++) is documented in https://www.tensorflow.org/lite/guide/ops_custom.

This PR makes a lot of changes to the TFLite kernel, which we cannot approve. Please follow the above guide and if there's any issue with it, please create a GitHub issue to follow up.

Thanks.

**Comment by mgokulkrish** [comment]
> Hi @mgokulkrish, the recommended way of adding custom op (only supported in C++) is documented in https://www.tensorflow.org/lite/guide/ops_custom.
> 
> This PR makes a lot of changes to the TFLite kernel, which we cannot approve. Please follow the above guide and if there's any issue with it, please create a GitHub issue to follow up.
> 
> Thanks.

Thanks for the review @JunyoungLim , based on the guide of custom op you shared, I think CPU side changes for the kernel should be fine. 

Maybe you could point out which part of GPU changes are wrong or where it is too big for kernel changes, I could learn from it.
Summary of my changes:-
1. added a transformations for adding two sub kernels to calculate mean and variance sum.
2. Added three kernels (tasks) for group norm, group mean sum and group var sum.
3. Added parser and selectors for it (as a custom layer, similar to MaxUnpooing2D which was done similarly before).

I have not modified any existing kernels in my changes.

Thanks again for reviewing the change, I would like to learn and make contribution in GPU delegate. So, a deep review will be a good learning and review process for me.

**Comment by gbaned** [comment]
Hi @sirakiin Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @sirakiin Can you please review this PR ? Thank you!

**Comment by ZTMIDGO** [comment]
> Hi @sirakiin Can you please review this PR ? Thank you!

@gbaned Hello, do you have the release aar of this pr? I tried to custom build the release, but it failed. I am trying to develop the Android version of stable diffusion. This is the Android version of stable diffusion I made using onnxruntime https://github.com/ZTMIDGO/Android-Stable-diffusion-ONNX . I hope to develop the Android version of TensorFlow.

**Comment by gbaned** [comment]
Hi @sirakiin Can you please assist on above comments from @ZTMIDGO. Thank you!

**Comment by gbaned** [comment]
Hi @JunyoungLim Can you please assist on above comments from @ZTMIDGO. Thank you!

**Comment by gbaned** [comment]
Hi @JunyoungLim Any update on this PR? Please. Thank you!

**Comment by gbaned** [comment]
Hi @JunyoungLim Any update on this PR? Please. Thank you!

**Comment by gbaned** [comment]
Hi @JunyoungLim Any update on this PR? Please. Thank you!

**Comment by gbaned** [comment]
Hi @JunyoungLim Any update on this PR? Please. Thank you!

**Comment by gbaned** [comment]
Hi @JunyoungLim Any update on this PR? Please. Thank you!

**Comment by gbaned** [comment]
Hi @sirakiin Any update on this PR? Please. Thank you!

**Comment by gbaned** [comment]
Hi @sirakiin Any update on this PR? Please. Thank you!

**Comment by keerthanakadiri** [comment]
Hi @sirakiin Any update on this PR? Please. Thank you!

**Comment by keerthanakadiri** [comment]
Hi @sirakiin , Any update on this PR? Please. Thank you!

**Comment by keerthanakadiri** [comment]
Hi @sirakiin , Any update on this PR? Please. Thank you!

**Comment by keerthanakadiri** [comment]
Hi @sirakiin , Any update on this PR? Please. Thank you!

**Comment by keerthanakadiri** [comment]
Hi @sirakiin , Any update on this PR? Please. Thank you!

**Comment by keerthanakadiri** [comment]
Hi @sirakiin , Any update on this PR? Please. Thank you!

**Comment by keerthanakadiri** [comment]
Hi @sirakiin , Any update on this PR? Please. Thank you!

**Comment by keerthanakadiri** [comment]
Hi @sirakiin , Any update on this PR? Please. Thank you!

**Comment by keerthanakadiri** [comment]
Hi @sirakiin , Any update on this PR? Please. Thank you!
